File: packages/gbnf-experiment/data/20260909T182714Z_2643e4be/ported_implementation/src/grammar-graph/pointers.ts
import type { GraphPointer } from "./graph-pointer.js";

export class Pointers {
  #pointers: Map<string, GraphPointer> = new Map();

  constructor(...pointers: GraphPointer[]) {
    for (const pointer of pointers) {
      this.add(pointer);
    }
  }

  add(pointer: GraphPointer): void {
    this.#pointers.set(pointer.id, pointer);
  }

  *[Symbol.iterator](): Generator<GraphPointer> {
    yield* this.#pointers.values();
  }

  get size(): number {
    return this.#pointers.size;
  }

  toString(): string {
    return `<Pointers [${[...this].map((p) => p.node.rule).join(", ")}]>`;
  }
}

File: proposed-projects/round-trip-experiment/reverse/20260909T124157Z_4ee67ba2/20260910T071826Z_7d46923a/ported_implementation/gbnf/grammar_graph/pointers.py
from typing import Dict, Iterator

from .graph_pointer import GraphPointer


class Pointers:
    def __init__(self, *pointers: GraphPointer):
        self.__pointers: Dict[str, GraphPointer] = {}
        for pointer in pointers:
            self.add(pointer)

    def add(self, pointer: GraphPointer) -> None:
        self.__pointers[pointer.id] = pointer

    def __iter__(self) -> Iterator[GraphPointer]:
        return iter(list(self.__pointers.values()))

    def __len__(self) -> int:
        return len(self.__pointers)

    def __str__(self) -> str:
        return f"<Pointers {', '.join(str(p.node.rule) for p in self)}>"

    def __repr__(self) -> str:
        return str(self)
